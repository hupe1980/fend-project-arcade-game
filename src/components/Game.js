import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { compose } from 'redux';
import Modal from 'react-modal';
import styled from 'styled-components';

import { Button, Heading, Image, Text, withAnimationFrame } from './common';
import withGame from './withGame';
import { createGame } from '../game';

Modal.setAppElement('#root');

const Container = styled.section``;

const GameOverContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`;

export function GameOver({ avatar, lives, level, points, onRestart }) {
  return (
    <Modal id="game-over" contentLabel="Game Over" isOpen={lives === 0}>
      <GameOverContainer>
        <Heading>!!! Game Over !!!</Heading>
        <Image src={`./${avatar}`} alt="avatar" />
        <Heading>Your stats:</Heading>
        <Text>{`Level: ${level}`}</Text>
        <Text>{`Points: ${points}`}</Text>
        <Button onClick={onRestart}>Restart Game</Button>
      </GameOverContainer>
    </Modal>
  );
}

GameOver.propTypes = {
  avatar: PropTypes.string.isRequired,
  lives: PropTypes.number.isRequired,
  level: PropTypes.number.isRequired,
  points: PropTypes.number.isRequired,
  onRestart: PropTypes.func.isRequired,
};

export class Game extends Component {
  constructor(props) {
    super(props);

    this.onRestart = this.onRestart.bind(this);
  }

  componentDidMount() {
    this.game = createGame(this.refs.canvas, this.props.changeState);
  }

  componentWillReceiveProps(nextProps) {
    this.game.changeAvatar(nextProps.avatar);
  }

  onAnimationFrame(time) {
    this.game.main(time);
  }

  onRestart() {
    this.game.restart();
  }

  render() {
    const { width, height, avatar, lives, level, points } = this.props;

    return (
      <Container>
        <canvas ref="canvas" width={width} height={height} />
        <GameOver
          avatar={avatar}
          lives={lives}
          level={level}
          points={points}
          onRestart={this.onRestart}
        />
      </Container>
    );
  }
}

Game.propTypes = {
  avatar: PropTypes.string.isRequired,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  lives: PropTypes.number.isRequired,
  level: PropTypes.number.isRequired,
  points: PropTypes.number.isRequired,
  changeState: PropTypes.func.isRequired,
};

export default compose(withGame, withAnimationFrame)(Game);
