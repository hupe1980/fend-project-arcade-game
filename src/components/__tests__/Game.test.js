import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import { Game, GameOver } from '../Game';

describe('Game', () => {
  it('renders correctly', () => {
    const wrapper = mount(
      <Game
        width={500}
        height={500}
        level={999}
        avatar={'image_url'}
        points={999}
        lives={999}
        changeState={i => i}
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe('GameOver', () => {
    it('renders correctly', () => {
      const wrapper = shallow(
        <GameOver
          level={999}
          avatar={'image_url'}
          points={999}
          lives={999}
          onRestart={i => i}
        />,
      );
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
});
