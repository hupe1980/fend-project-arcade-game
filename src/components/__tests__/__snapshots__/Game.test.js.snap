// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game GameOver renders correctly 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  closeTimeoutMS={0}
  contentLabel="Game Over"
  id="game-over"
  isOpen={false}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <styled.div>
    <styled.h1>
      !!! Game Over !!!
    </styled.h1>
    <styled.img
      alt="avatar"
      src="./image_url"
    />
    <styled.h1>
      Your stats:
    </styled.h1>
    <styled.p>
      Level: 999
    </styled.p>
    <styled.p>
      Points: 999
    </styled.p>
    <styled.button
      onClick={[Function]}
    >
      Restart Game
    </styled.button>
  </styled.div>
</Modal>
`;

exports[`Game renders correctly 1`] = `
<Game
  avatar="image_url"
  changeState={[Function]}
  height={500}
  level={999}
  lives={999}
  points={999}
  width={500}
>
  <styled.section>
    <section
      className="sc-bZQynM bOrPFB"
    >
      <canvas
        height={500}
        width={500}
      />
      <GameOver
        avatar="image_url"
        level={999}
        lives={999}
        onRestart={[Function]}
        points={999}
      >
        <Modal
          ariaHideApp={true}
          bodyOpenClassName="ReactModal__Body--open"
          closeTimeoutMS={0}
          contentLabel="Game Over"
          id="game-over"
          isOpen={false}
          parentSelector={[Function]}
          portalClassName="ReactModalPortal"
          shouldCloseOnEsc={true}
          shouldCloseOnOverlayClick={true}
          shouldFocusAfterRender={true}
          shouldReturnFocusAfterClose={true}
        >
          <ModalPortal
            ariaHideApp={true}
            bodyOpenClassName="ReactModal__Body--open"
            closeTimeoutMS={0}
            contentLabel="Game Over"
            defaultStyles={
              Object {
                "content": Object {
                  "WebkitOverflowScrolling": "touch",
                  "background": "#fff",
                  "border": "1px solid #ccc",
                  "borderRadius": "4px",
                  "bottom": "40px",
                  "left": "40px",
                  "outline": "none",
                  "overflow": "auto",
                  "padding": "20px",
                  "position": "absolute",
                  "right": "40px",
                  "top": "40px",
                },
                "overlay": Object {
                  "backgroundColor": "rgba(255, 255, 255, 0.75)",
                  "bottom": 0,
                  "left": 0,
                  "position": "fixed",
                  "right": 0,
                  "top": 0,
                },
              }
            }
            id="game-over"
            isOpen={false}
            parentSelector={[Function]}
            portalClassName="ReactModalPortal"
            shouldCloseOnEsc={true}
            shouldCloseOnOverlayClick={true}
            shouldFocusAfterRender={true}
            shouldReturnFocusAfterClose={true}
            style={
              Object {
                "content": Object {},
                "overlay": Object {},
              }
            }
          />
        </Modal>
      </GameOver>
    </section>
  </styled.section>
</Game>
`;
